pkg/webhooks/controlplanemachineset/webhooks.go

```diff
@@ -62,15 +86,179 @@ var _ webhook.CustomValidator = &ControlPlaneMachineSetWebhook{}
+	}
+
+	// Ensure Control Plane Machine count matches the ControlPlaneMachineSet replicas
+	if cpms.Spec.Replicas == nil {
+		return errReplicasNil
+	}
+
+	if int(*cpms.Spec.Replicas) != len(controlPlaneMachines) {
+		return fmt.Errorf("%w: replicas: %v, control plane machines: %v", errControlPlaneReplicasDoesNotMatchMachineCount, *cpms.Spec.Replicas, len(controlPlaneMachines))
+	}
+
+	// Ensure CPMS created with invalid name is not allowed
+	if cpms.Name != "cluster" {
+		return fmt.Errorf("%w: invalid name %s", errControlPlaneInvalidName, cpms.Name)
+	}
+
+	// Ensure required labels are set and all machines are matching the label selector
+	if err := checkMachineLabels(cpms); err != nil {
+		return fmt.Errorf("invalid label selector configuration: %w", err)
+	}
+
+	// Ensure failure domains of Control Plane Machines match the ControlPlaneMachineSet on create
+	switch cpms.Spec.Template.MachineType {
+	case machinev1.OpenShiftMachineV1Beta1MachineType:
+		if err := checkFailureDomains(cpms, controlPlaneMachines); err != nil {
+			return fmt.Errorf("invalid failure domain configuration: %w", err)
+		}
+	default:
+		return errUnknownMachineTemplateType
+	}
+
 	return nil
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
 func (r *ControlPlaneMachineSetWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	if oldObj == nil {
+		return errUpdateNilCPMS
+	}
+
+	oldCPMS, ok := oldObj.(*machinev1.ControlPlaneMachineSet)
+	if !ok {
+		return errTypeMismatchCondition
+	}
+
+	newCPMS, ok := newObj.(*machinev1.ControlPlaneMachineSet)
+	if !ok {
+		return errTypeMismatchCondition
+	}
+
+	// Ensure spec.Replicas is immutable on update
+	if oldCPMS.Spec.Replicas == nil || newCPMS.Spec.Replicas == nil {
+		return errReplicasNil
+	}
+
+	if *oldCPMS.Spec.Replicas != *newCPMS.Spec.Replicas {
+		return errControlPlaneReplicasChange
+	}
+
+	// Ensure selector is immutable on update
+	if !reflect.DeepEqual(oldCPMS.Spec.Selector, newCPMS.Spec.Selector) {
+		return errControlPlaneLabelSelectorChange
+	}
+
+	// Ensure required labels are set and all machines are matching the label selector
+	if err := checkMachineLabels(newCPMS); err != nil {
+		return fmt.Errorf("invalid label selector configuration: %w", err)
+	}
+
 	return nil
 }
 
 // ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
 func (r *ControlPlaneMachineSetWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) error {
 	return nil
 }
+
+// fetchControlPlaneMachines returns all control plane machines in the cluster.
+func (r *ControlPlaneMachineSetWebhook) fetchControlPlaneMachines(ctx context.Context) ([]*machinev1beta1.Machine, error) {
+	machineList := &machinev1beta1.MachineList{}
+	if err := r.client.List(ctx, machineList); err != nil {
+		return nil, fmt.Errorf("error querying api for machineList object: %w", err)
```

Comment:
Nit, prefer this wording
```suggestion
		return nil, fmt.Errorf("error querying api for machines: %w", err)
```